fix: implement contact form validation and checkbox reset

alisa-korniienko opened the pull request:
## Description

This PR fixes validation issues in the contact form according to specifications. Implements comprehensive field validation, success confirmation modal, and proper form reset including checkbox state.
Includes a custom `useAutoHideMessage` hook that manages temporary message display with automatic hiding after a specified timeout, used for max-length validation messages.

## Type of change

- [ ] New feature
- [x] Bug fix
- [ ] Documentation update
- [ ] Refactoring / Tech debt
- [ ] CI/CD improvement

#### Validation Improvements:
- **Name field**:
  - Added `.trim()` to prevent whitespace-only input
  - Min length: 2 characters
  - Max length: 50 characters (with input blocking)
  - Auto-hiding max length message (3 seconds)
  
- **Message field**:
  - Added `.trim()` to prevent whitespace-only input
  - Min length: 10 characters
  - Max length: 1000 characters (with input blocking)
  - Auto-hiding max length message (3 seconds)

#### UX Enhancements
- Added success confirmation modal after form submission
- Form clears after closing the confirmation modal
- Max length messages appear in red when limit is reached, then auto-hide after 3 seconds
- Input is blocked at maximum length to prevent invalid submissions

#### Localization
- Added validation error messages in Ukrainian (`uk.json`)
- Added validation error messages in English (`en.json`)
- Added confirmation modal text for both languages

### Files Changed
- `app/shared/components/forms/contact-form/ContactForm.tsx`
- `i18n/messages/uk.json`
- `i18n/messages/en.json`
- `app/shared/hooks/use-auto-hide-message/useAutoHideMessage.ts`
- `app/shared/hooks/use-auto-hide-message/useAutoHideMessage.test.ts`

## How to test

Navigate to `/contacts` page.

### Test Scenarios:

**1. Empty Form Validation**
- Click "Надіслати запит" without filling fields
- Expected: All required fields show validation errors

**2. Name Field Min/Max Length**
- Enter "A" - Error: "Імʼя має містити щонайменше 2 символи"
- Enter "AB" - Error disappears
- Type 50 characters - Red message appears, auto-hides after 3s
- Try typing 51st character - Blocked

**3. Whitespace Validation**
- Enter only spaces "   " in Name or Message - Error: "Поле обов'язкове для заповнення"

**4. Email Validation**
- Leave empty - Error: "Будь ласка, вкажіть вашу електронну адресу"
- Enter "invalid" - Error: "Введіть коректну email-адресу"
- Enter "test@example.com" - Valid

**5. Message Field Min/Max Length**
- Enter "Test" (4 chars) - Error: "Коментар має містити щонайменше 10 символів"
- Enter 10+ characters - Valid
- Type 1000 characters - Red message appears, auto-hides after 3s
- Try typing 1001st character - Blocked

**6. Policy Checkbox**
- Submit without checking - Error: "Щоб продовжити, потрібно дати згоду"
- Check checkbox - Error disappears

**7. Successful Submission & Form Reset**
- Fill valid data: Name, Email, Message (10+ chars), check Policy
- Click "Надіслати запит"
- Expected: Modal appears with "ДяКуЮ Ваш запит надіслано"
- Click "Закрити"
- **Expected: all fields cleared including checkbox unchecked** 

**8. Form Submittable at Max Length**
- Fill Name with 50 characters (max message visible)
- Fill other fields correctly
- Click "Надіслати запит"
- Expected: Form submits successfully



## Video / Screenshots


https://github.com/user-attachments/assets/3e24242b-5639-41d8-abee-56a2a0ac7e18


## Checklist

- [ ] Empty fields show required field errors
- [ ] Name field with < 2 characters shows min length error
- [ ] Name field at 50 characters blocks 51st character input
- [ ] Name field shows max length error that auto-hides after 3 seconds
- [ ] Message field at 1000 characters blocks 1001st character input
- [ ] Message field shows max length error that auto-hides after 3 seconds
- [ ] Whitespace-only input is rejected due to `.trim()`
- [ ] Success modal appears after valid form submission
- [ ] Form clears after closing success modal

---

alisa-korniienko committed:
fix: implement contact form validation according to specifications (#999)

- Add trim validation for name and message fields
- Implement max length validation (name: 50 chars, message: 1000 chars)
- Add input blocking at max length with auto-hiding error messages
- Add success confirmation modal after form submission
- Add localization keys for validation errors and confirmation (uk/en)
- Refactor: extract timeout constant, improve code organization

alisa-korniienko committed:
fix: implement contact form validation and checkbox reset (#999)

- Add trim validation for name and message fields
- Implement max length validation (name: 50, message: 1000)
- Add input blocking at max length with auto-hiding messages (3s)
- Add success confirmation modal after form submission
- Fix checkbox not clearing after form reset with controlled state
- Change policy validation from z.literal(true) to z.boolean().refine()
- Add localization keys for validation errors (uk/en)
- Refactor: extract AUTO_HIDE_MESSAGE_TIMEOUT constant

stsvt commented on app/shared/components/forms/contact-form/ContactForm.tsx:
```diff
@@ -42,107 +48,207 @@ function ContactForm({ onSubmit }: Readonly<ContactFormProps>) {
+    } else {
+      setShowMessageMaxMessage(false);
+    }
+  }, [isMessageAtMaxLength]);
```
This logic looks very similar for both cases. Maybe we could extract it into a custom hook to avoid duplication and keep the component simpler?

alisa-korniienko replied:
Makes sense! I kept it inline initially since it was component-specific, but extracting it into a custom hook definitely improves maintainability. Refactored into `useAutoHideMessage` with tests

stsvt replied:
Great work! Thank you)

alisa-korniienko committed:
refactor: extract auto-hide message logic into custom hook

- Create useAutoHideMessage hook to eliminate duplicate useEffect code
- Add comprehensive tests for the new hook
- Simplify ContactForm component by removing 2 useState and 2 useEffect
- Improve code reusability and maintainability
